File: packages/less/src/less/tree/combinator.test.js
import { describe, it, expect, vi } from 'vitest';
import Combinator from './combinator';

describe('Combinator', () => {
    describe('constructor', () => {
        it('should create a combinator with empty value', () => {
            const combinator = new Combinator();
            expect(combinator.value).toBe('');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });

        it('should create a combinator with space value', () => {
            const combinator = new Combinator(' ');
            expect(combinator.value).toBe(' ');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });

        it('should create a combinator with trimmed value', () => {
            const combinator = new Combinator('  >  ');
            expect(combinator.value).toBe('>');
            expect(combinator.emptyOrWhitespace).toBe(false);
        });

        it('should handle null value', () => {
            const combinator = new Combinator(null);
            expect(combinator.value).toBe('');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });

        it('should handle undefined value', () => {
            const combinator = new Combinator(undefined);
            expect(combinator.value).toBe('');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });

        it('should handle empty string value', () => {
            const combinator = new Combinator('');
            expect(combinator.value).toBe('');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });

        it('should handle various whitespace combinations', () => {
            const combinator = new Combinator('\t\n\r  >  \t\n\r');
            expect(combinator.value).toBe('>');
            expect(combinator.emptyOrWhitespace).toBe(false);
        });

        it('should handle special characters in input', () => {
            const combinator = new Combinator('  *  ');
            expect(combinator.value).toBe('*');
            expect(combinator.emptyOrWhitespace).toBe(false);
        });

        it('should handle unicode whitespace characters', () => {
            const combinator = new Combinator(
                '\u00A0\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000\ufeff>'
            );
            expect(combinator.value).not.toBe('');
            expect(combinator.value).toContain('>');
            expect(combinator.emptyOrWhitespace).toBe(false);
        });

        it('should handle very long strings with multiple spaces', () => {
            const combinator = new Combinator('   ' + '>'.repeat(100) + '   ');
            expect(combinator.value).toBe('>'.repeat(100));
            expect(combinator.emptyOrWhitespace).toBe(false);
        });

        it('should handle strings containing only whitespace', () => {
            const combinator = new Combinator('\t\n\r ');
            expect(combinator.value).toBe('');
            expect(combinator.emptyOrWhitespace).toBe(true);
        });
    });

    describe('genCSS', () => {
        it('should generate CSS with no spaces for empty combinator', () => {
            const combinator = new Combinator('');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith('');
        });

        it('should generate CSS with no spaces for space combinator', () => {
            const combinator = new Combinator(' ');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith(' ');
        });

        it('should generate CSS with spaces for non-empty combinator when not compressed', () => {
            const combinator = new Combinator('>');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith(' > ');
        });

        it('should generate CSS without spaces for non-empty combinator when compressed', () => {
            const combinator = new Combinator('>');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: true }, output);
            expect(output.add).toHaveBeenCalledWith('>');
        });

        it('should generate CSS without spaces for pipe combinator even when not compressed', () => {
            const combinator = new Combinator('|');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith('|');
        });

        it('should handle other common combinators', () => {
            const combinators = ['+', '~', '>>'];
            const output = { add: vi.fn() };

            combinators.forEach((value) => {
                const combinator = new Combinator(value);
                combinator.genCSS({ compress: false }, output);
                expect(output.add).toHaveBeenCalledWith(` ${value} `);
            });
        });

        it('should handle multiple spaces in input', () => {
            const combinator = new Combinator('  >  ');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith(' > ');
        });

        it('should handle special characters in combinator value', () => {
            const combinator = new Combinator('*');
            const output = { add: vi.fn() };
            combinator.genCSS({ compress: false }, output);
            expect(output.add).toHaveBeenCalledWith(' * ');
        });

        it('should handle all special no-space combinators', () => {
            const specialCombinators = ['', ' ', '|'];
            const output = { add: vi.fn() };

            specialCombinators.forEach((value) => {
                const combinator = new Combinator(value);
                combinator.genCSS({ compress: false }, output);
                expect(output.add).toHaveBeenCalledWith(value);
            });
        });

        it('should handle undefined context', () => {
            const combinator = new Combinator('>');
            const output = { add: vi.fn() };
            expect(() => combinator.genCSS(undefined, output)).toThrow();
        });

        it('should handle null context', () => {
            const combinator = new Combinator('>');
            const output = { add: vi.fn() };
            expect(() => combinator.genCSS(null, output)).toThrow();
        });

        it('should handle context with undefined compress property', () => {
            const combinator = new Combinator('>');
            const output = { add: vi.fn() };
            combinator.genCSS({}, output);
            expect(output.add).toHaveBeenCalledWith(' > ');
        });

        it('should handle more special combinator characters', () => {
            const specialChars = [
                '/',
                '\\',
                ':',
                ';',
                '!',
                '@',
                '#',
                '$',
                '%',
                '^',
                '&',
                '=',
                '?'
            ];
            const output = { add: vi.fn() };

            specialChars.forEach((value) => {
                const combinator = new Combinator(value);
                combinator.genCSS({ compress: false }, output);
                expect(output.add).toHaveBeenCalledWith(` ${value} `);
            });
        });
    });

    describe('inheritance from Node', () => {
        it('should have Node properties', () => {
            const combinator = new Combinator('>');
            expect(combinator.parent).toBeNull();
            expect(combinator.visibilityBlocks).toBeUndefined();
            expect(combinator.nodeVisible).toBeUndefined();
            expect(combinator.rootNode).toBeNull();
            expect(combinator.parsed).toBeNull();
        });

        it('should be able to set parent', () => {
            const combinator = new Combinator('>');
            const parent = {};
            combinator.setParent(combinator, parent);
            expect(combinator.parent).toBe(parent);
        });

        it('should be able to set parent for array of nodes', () => {
            const combinator1 = new Combinator('>');
            const combinator2 = new Combinator('+');
            const parent = {};
            combinator1.setParent([combinator1, combinator2], parent);
            expect(combinator1.parent).toBe(parent);
            expect(combinator2.parent).toBe(parent);
        });
    });
});

File: packages/less/src/less/tree/combinator_test.go
package tree

import (
	"strings"
	"testing"
)

func TestCombinator(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("should create a combinator with empty value", func(t *testing.T) {
			combinator := NewCombinator("")
			if combinator.Value != "" {
				t.Errorf("Expected empty value, got %s", combinator.Value)
			}
			if !combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be true")
			}
		})

		t.Run("should create a combinator with space value", func(t *testing.T) {
			combinator := NewCombinator(" ")
			if combinator.Value != " " {
				t.Errorf("Expected space value, got %s", combinator.Value)
			}
			if !combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be true")
			}
		})

		t.Run("should create a combinator with trimmed value", func(t *testing.T) {
			combinator := NewCombinator("  >  ")
			if combinator.Value != ">" {
				t.Errorf("Expected '>', got %s", combinator.Value)
			}
			if combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be false")
			}
		})

		t.Run("should handle empty string value", func(t *testing.T) {
			combinator := NewCombinator("")
			if combinator.Value != "" {
				t.Errorf("Expected empty value, got %s", combinator.Value)
			}
			if !combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be true")
			}
		})

		t.Run("should handle various whitespace combinations", func(t *testing.T) {
			combinator := NewCombinator("\t\n\r  >  \t\n\r")
			if combinator.Value != ">" {
				t.Errorf("Expected '>', got %s", combinator.Value)
			}
			if combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be false")
			}
		})

		t.Run("should handle special characters in input", func(t *testing.T) {
			combinator := NewCombinator("  *  ")
			if combinator.Value != "*" {
				t.Errorf("Expected '*', got %s", combinator.Value)
			}
			if combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be false")
			}
		})

		t.Run("should handle strings containing only whitespace", func(t *testing.T) {
			combinator := NewCombinator("\t\n\r ")
			if combinator.Value != "" {
				t.Errorf("Expected empty value, got %s", combinator.Value)
			}
			if !combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be true")
			}
		})
	})

	t.Run("genCSS", func(t *testing.T) {
		t.Run("should generate CSS with no spaces for empty combinator", func(t *testing.T) {
			combinator := NewCombinator("")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != "" {
				t.Errorf("Expected empty string, got %v", output)
			}
		})

		t.Run("should generate CSS with no spaces for space combinator", func(t *testing.T) {
			combinator := NewCombinator(" ")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != " " {
				t.Errorf("Expected space, got %v", output)
			}
		})

		t.Run("should generate CSS with spaces for non-empty combinator when not compressed", func(t *testing.T) {
			combinator := NewCombinator(">")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != " > " {
				t.Errorf("Expected ' > ', got %v", output)
			}
		})

		t.Run("should generate CSS without spaces for non-empty combinator when compressed", func(t *testing.T) {
			combinator := NewCombinator(">")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": true}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != ">" {
				t.Errorf("Expected '>', got %v", output)
			}
		})

		t.Run("should generate CSS without spaces for pipe combinator even when not compressed", func(t *testing.T) {
			combinator := NewCombinator("|")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != "|" {
				t.Errorf("Expected '|', got %v", output)
			}
		})

		t.Run("should handle other common combinators", func(t *testing.T) {
			combinators := []string{"+", "~", ">>"}
			for _, value := range combinators {
				combinator := NewCombinator(value)
				var output []string
				outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
					output = append(output, chunk.(string))
				}
				combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
				expected := " " + value + " "
				if len(output) != 1 || output[0] != expected {
					t.Errorf("Expected '%s', got %v", expected, output)
				}
			}
		})

		t.Run("should handle special characters in combinator value", func(t *testing.T) {
			combinator := NewCombinator("*")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != " * " {
				t.Errorf("Expected ' * ', got %v", output)
			}
		})

		t.Run("should handle all special no-space combinators", func(t *testing.T) {
			specialCombinators := []string{"", " ", "|"}
			for _, value := range specialCombinators {
				combinator := NewCombinator(value)
				var output []string
				outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
					output = append(output, chunk.(string))
				}
				combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
				if len(output) != 1 || output[0] != value {
					t.Errorf("Expected '%s', got %v", value, output)
				}
			}
		})

		t.Run("should handle unicode whitespace characters", func(t *testing.T) {
			combinator := NewCombinator("\u00A0\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000\ufeff>")
			if combinator.Value != ">" {
				t.Errorf("Expected '>', got %s", combinator.Value)
			}
			if combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be false")
			}
		})

		t.Run("should handle very long strings with multiple spaces", func(t *testing.T) {
			longString := "   " + strings.Repeat(">", 100) + "   "
			combinator := NewCombinator(longString)
			if combinator.Value != strings.Repeat(">", 100) {
				t.Error("Expected long string of '>' characters")
			}
			if combinator.EmptyOrWhitespace {
				t.Error("Expected EmptyOrWhitespace to be false")
			}
		})

		t.Run("should handle more special combinator characters", func(t *testing.T) {
			specialChars := []string{"/", "\\", ":", ";", "!", "@", "#", "$", "%", "^", "&", "=", "?"}
			for _, value := range specialChars {
				combinator := NewCombinator(value)
				var output []string
				outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
					output = append(output, chunk.(string))
				}
				combinator.GenCSS(map[string]interface{}{"compress": false}, &CSSOutput{Add: outputFunc})
				expected := " " + value + " "
				if len(output) != 1 || output[0] != expected {
					t.Errorf("Expected '%s', got %v", expected, output)
				}
			}
		})

		t.Run("should handle undefined context", func(t *testing.T) {
			combinator := NewCombinator(">")
			var output []string
			outputFunc := func(chunk interface{}, fileInfo interface{}, index interface{}) {
				output = append(output, chunk.(string))
			}
			combinator.GenCSS(nil, &CSSOutput{Add: outputFunc})
			if len(output) != 1 || output[0] != " > " {
				t.Errorf("Expected ' > ', got %v", output)
			}
		})
	})

	t.Run("inheritance from Node", func(t *testing.T) {
		t.Run("should have Node properties", func(t *testing.T) {
			combinator := NewCombinator(">")
			if combinator.Parent != nil {
				t.Error("Expected Parent to be nil")
			}
			if combinator.VisibilityBlocks != nil {
				t.Error("Expected VisibilityBlocks to be nil")
			}
			if combinator.NodeVisible != nil {
				t.Error("Expected NodeVisible to be nil")
			}
			if combinator.RootNode != nil {
				t.Error("Expected RootNode to be nil")
			}
			if combinator.Parsed != nil {
				t.Error("Expected Parsed to be nil")
			}
		})

		t.Run("should be able to set parent", func(t *testing.T) {
			combinator := NewCombinator(">")
			parent := NewNode()
			combinator.SetParent(combinator.Node, parent)
			if combinator.Parent != parent {
				t.Error("Parent was not set correctly")
			}
		})

		t.Run("should be able to set parent for array of nodes", func(t *testing.T) {
			combinator1 := NewCombinator(">")
			combinator2 := NewCombinator("+")
			parent := NewNode()
			combinator1.SetParent([]*Node{combinator1.Node, combinator2.Node}, parent)
			if combinator1.Parent != parent {
				t.Error("Parent was not set correctly for first combinator")
			}
			if combinator2.Parent != parent {
				t.Error("Parent was not set correctly for second combinator")
			}
		})
	})
} 